src/rgw/cls_fifo_legacy.cc

```diff
@@ -825,11 +825,17 @@ int FIFO::_prepare_new_head(const DoutPrefixProvider *dpp, std::uint64_t tid, op
   ldpp_dout(dpp, 20) << __PRETTY_FUNCTION__ << ":" << __LINE__
 		 << " entering: tid=" << tid << dendl;
   std::unique_lock l(m);
+  preparing = true;
   std::int64_t new_head_num = info.head_part_num + 1;
   auto max_push_part_num = info.max_push_part_num;
   auto version = info.version;
   l.unlock();
-
+  auto sg(make_scope_guard(
```

Comment:
Do we need to remove this one? Otherwise, we are declaring that preparation of a new head has completed but we haven't updated/set the new head yet? Thus, allowing another thread calling into `_prepare_new_head`.

Reply:
Is the suggestion that we should hold `preparing` true until we process the journal?

Reply:
Thinking about it, I don't think we need to keep `preparing` true any longer than is needed to call `update_meta` so unless there' something I'm missing, I think the scope guard is fine.

What we should do is set 'preparing' under the lock in need_new_head and have `_prepare_new_head` require it to be set when it's called.